crates/bevy_ecs/src/schedule/ambiguity_detection.rs

```diff
@@ -0,0 +1,284 @@
+        for dependant in all_dependants[index].ones() {
+            dependants.union_with(&all_dependants[dependant]);
+            dependants.insert(dependant);
+        }
+        all_dependants[index] = dependants;
+    }
+    let mut all_relations = all_dependencies
+        .drain(..)
+        .zip(all_dependants.drain(..))
+        .enumerate()
+        .map(|(index, (dependencies, dependants))| {
+            let mut relations = FixedBitSet::with_capacity(systems.len());
+            relations.union_with(&dependencies);
+            relations.union_with(&dependants);
+            relations.insert(index);
+            relations
+        })
+        .collect::<Vec<FixedBitSet>>();
+    let mut ambiguities = Vec::new();
+    let full_bitset: FixedBitSet = (0..systems.len()).collect();
+    let mut processed = FixedBitSet::with_capacity(systems.len());
+    for (index_a, relations) in all_relations.drain(..).enumerate() {
+        // TODO: prove that `.take(index_a)` would be correct here, and uncomment it if so.
+        for index_b in full_bitset.difference(&relations)
+        // .take(index_a)
+        {
+            if !processed.contains(index_b)
+                && all_ambiguity_sets[index_a].is_disjoint(&all_ambiguity_sets[index_b])
+                && !should_ignore_ambiguity(systems, index_a, index_b, crates_filter)
+            {
+                let a_access = systems[index_a].component_access();
+                let b_access = systems[index_b].component_access();
+                if let (Some(a), Some(b)) = (a_access, b_access) {
+                    let conflicts = a.get_conflicts(b);
+                    if !conflicts.is_empty() {
+                        ambiguities.push((index_a, index_b, conflicts));
+                    }
+                } else {
+                    ambiguities.push((index_a, index_b, Vec::new()));
+                }
+            }
+        }
+        processed.insert(index_a);
+    }
+    ambiguities
+}
+
+impl ReportExecutionOrderAmbiguities {
+    /// Disables all messages reported by the ambiguity checker.
+    pub fn off() -> Self {
+        Self {
+            level: AmbiguityReportLevel::Off,
+            ..Default::default()
+        }
+    }
+
+    /// Displays only the number of unresolved ambiguities detected by the ambiguity checker. This is the default behavior.
+    pub fn minimal() -> Self {
+        Self {
+            level: AmbiguityReportLevel::Minimal,
+            ..Default::default()
+        }
+    }
+
+    /// Displays a full report of ambiguities detected by the ambiguity checker.
+    pub fn verbose() -> Self {
+        Self {
+            level: AmbiguityReportLevel::Verbose,
+            ..Default::default()
+        }
+    }
+
+    /// Adds the given crate to be ignored by ambiguity checker. Check [`ReportExecutionOrderAmbiguities`] for more details.
+    pub fn ignore(mut self, crate_prefix: &str) -> Self {
```

Comment:
Would it be better to do something like:
```suggestion
    pub fn ignore<I: Into<String>>(mut self, crate_prefix: I) -> Self {
```
in order to support other types?

Reply:
I generally am not thrilled with this design: I think it deserves to be a field on the Plugin trait.

As a result, I don't think it's worth improving it in this way now.